ARCHITECTURE.md

```diff
@@ -0,0 +1,517 @@
+The protobuf definitions can be found in `internal/server/proto/server.proto`.
+We use one giant protobuf file to avoid some of the pitfalls of a multi-file
+protobuf setup. The protobuf definitions are compiled during build-time for the server.
+This is represented by the dashed line, meaning we don't do this at runtime.
+
+The Go package paths that implement each component are noted alongside 
+the package. 
+
+For data storage in the single process implementation, we use BoltDB for
+persistent storage with MemDB for in-memory indexing and storage. On server
+startup, we build the MemDB database; we do not persist any part of the MemDB
+data so as the database increases, startup times will also increase but even
+large databases in testing have started in a few seconds.
+
+This arrangement may seem strange, but it is the same set of technology
+choices that back other high-scale software at HashiCorp: Consul, Nomad,
+and Vault. We are familiar with it and we know how to make it scale. 
+Additionally, it allows us to have a single-binary setup without any
+external dependencies (such as a SQL server). We plan on likely implementing
+a SQL backend for a multi-process-capable server implementation in the future.
+
+### Job System
+
+Waypoint uses a [task queue system](https://en.wikipedia.org/wiki/Scheduling_(computing)#task_queue)
+internally to represent most operations such as builds, deploys, etc. The
+Waypoint Server queues jobs, and Waypoint Runners dequeue and execute the
+queued jobs. The Waypoint Server process itself never executes a job.
+
+#### State Machine
+
+A job is represented in Waypoint as a state machine. The diagram below shows
+the full state machine a job:
+
+<p align="center"><img src=".github/images/job-states.png" width="70%"></p>
+
+A job begins **QUEUED** as soon as it is enqueued using the `QueueJob` API. 
+Additionally, the Waypoint server sometimes self-queues jobs such as for
+polling. 
+
+When a Runner process starts or has capacity, it can request a job from
+the Waypoint server. If no jobs are available, this request will block and
+the runner will wait. If a job is available, the Waypoint server assigns the
+job and the job enters the **WAITING** state. During this state, we're waiting
+for the runner to acknowledge that it has taken responsibility for the job.
+
+During the **QUEUED** and **WAITING** states, a job may expire. A job may optionally
+have an expiration deadline associated with it. If a job has a deadline, then at
+that time, the job will immediately move to the **ERROR** state with a message noting
+it expired.
+
+While a job is in the **WAITING**, it will not be assigned to any other runner.
+
+After being assigned a job, the runner acknowledges (acks) the job. This means that
+the runner has taken responsibility for the job and the job enters the **RUNNING**
+state. If the runner nacks (negative acknowledgement) the job or the runner doesn't
```

Comment:
```suggestion
state. If the runner negatively acknowledges (nacks) the job or the runner doesn't
```

I feel this improves readability by putting both the short terms (ack, nack) in parenthesis, instead of being opposite.